File: Nov2020/30Nov/CrushingViolence.cpp
/*
Problem Name: Crushing Violence
Problem Link: https://www.hackerearth.com/practice/data-structures/hash-tables/basics-of-hash-tables/practice-problems/algorithm/crushing-violence/
*/

#include<bits/stdc++.h>
using namespace std;

#define test(t)     int t; cin>>t; while(t--)

int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    test(t){
        int n;
        cin>>n;
        int boy[n+1]={0}, girl[n+1]={0}, boyhitby[n+1]={0}, girlhitby[n+1]={0}, worst = 0, pairs = 0;
        for(int i=1;i<=n;++i)
            cin>>boy[i];
        for(int i=1;i<=n;++i)
            cin>>girl[i];
    
        for(int i=0;i<n;++i){
            if(i+1 != girl[boy[i+1]]){
                boyhitby[girl[boy[i+1]]]++;
                girlhitby[boy[girl[i+1]]]++;

                worst = max(worst, boyhitby[girl[boy[i+1]]]);
                worst = max(worst, girlhitby[boy[girl[i+1]]]);

                if(i+1 == girl[boy[girl[boy[i+1]]]]){
                    pairs++;
                }
            } 
        }
        cout<<worst<<" "<<pairs<<"\n";
    }
    return 0;
}

File: Nov2020/30Nov/CrushingViolence.java
/*
Problem Name: Crushing Violence
Problem Link: https://www.hackerearth.com/practice/data-structures/hash-tables/basics-of-hash-tables/practice-problems/algorithm/crushing-violence/
*/
import java.io.*;
import java.util.*;
class TestClass {
    public static void main(String args[] ) throws Exception {
        Scanner in = new Scanner(System.in);
        PrintWriter out = new PrintWriter(System.out);
        int t = in.nextInt();
        for(;t>0;t--){
            int n = in.nextInt();
            int boy[] = new int[n], girl[] = new int[n];
            for(int i=0;i<n;i++)
                boy[i] = in.nextInt() - 1;
            for(int i=0;i<n;i++)
                girl[i] = in.nextInt() - 1;
            int boybeat[] = new int[n], girlbeat[] = new int[n];
            int pairs = 0;
            for(int i=0;i<n;i++){
                if(i != girl[boy[i]]){
                    boybeat[girl[boy[i]]]++;
                    if(girl[boy[girl[boy[i]]]] == i)
                        pairs++;
                }
                if(i != boy[girl[i]]){
                    girlbeat[boy[girl[i]]]++;
                    if(boy[girl[boy[girl[i]]]] == i)
                        pairs++;
                }
            }
            int max = 0;
            for(int i=0;i<n;i++)
                max = Math.max(max, Math.max(boybeat[i],girlbeat[i]));
            out.println(max+" "+(pairs/2));
        }
        out.flush();
    }
}
